src/python/pants/core/target_types_test.py

```diff
@@ -367,3 +372,110 @@ def gen_resource_tgt(rel_fp: str, tags: list[str] | None = None) -> ResourceTarg
+        "len=500,"
+        'sha256="8a5ee63e1b79ba2733e7ff4290b6eefea60e7f3a1ccb6bb519535aaf92b44967"'
+    )
+    rule_runner.write_files(
+        {
+            "assets/BUILD": dedent(
+                f"""\
+                {asset_type}(
+                    name='antigravity',
+                    source=http_source(
+                        {http_source_info},
+                    ),
+                )
+                {asset_type}(
+                    name='antigravity_renamed',
+                    source=http_source(
+                        {http_source_info},
+                        filename="antigravity_renamed.py",
+                    ),
+                )
+                """
+            ),
+            "app/app.py": textwrap.dedent(
+                """\
+                import pathlib
+
+                assets_path = pathlib.Path(__file__).parent.parent / "assets"
+                for path in assets_path.iterdir():
+                    print(path.name)
+                    assert "https://xkcd.com/353/" in path.read_text()
+                """
+            ),
+            "app/BUILD": textwrap.dedent(
+                """\
+                python_source(
+                    source="app.py",
+                    dependencies=[
+                        "assets:antigravity",
+                        "assets:antigravity_renamed",
+                    ]
+                )
+                pex_binary(name="app.py", entry_point='app.py')
+                """
+            ),
+        }
+    )
+    with mock_console(rule_runner.options_bootstrapper) as (console, stdout_reader):
+        rule_runner.run_goal_rule(
+            run.Run,
+            args=["app/app.py"],
+            env_inherit={"PATH", "PYENV_ROOT", "HOME"},
+        )
+        stdout = stdout_reader.get_stdout()
+        assert "antigravity.py" in stdout
+        assert "antigravity_renamed.py" in stdout
+
+
+@pytest.mark.parametrize(
+    "kwargs, exc_match",
+    [
+        (
+            dict(url=None, len=0, sha256=""),
+            pytest.raises(TypeError, match=r"`url` must be a `str`"),
+        ),
+        (
+            dict(url="http://foo/bar", len="", sha256=""),
+            pytest.raises(TypeError, match=r"`len` must be a `int`"),
+        ),
+        (
+            dict(url="http://foo/bar", len=0, sha256=123),
+            pytest.raises(TypeError, match=r"`sha256` must be a `str`"),
+        ),
+        (
+            dict(url="http://foo/bar", len=0, sha256="", filename=10),
+            pytest.raises(TypeError, match=r"`filename` must be a `str`"),
+        ),
+        (
+            dict(url="http://foo/bar/", len=0, sha256=""),
+            pytest.raises(ValueError, match=r"Couldn't deduce filename"),
+        ),
+        (
+            dict(url="http://foo/bar/", len=0, sha256="", filename="../foobar.txt"),
+            pytest.raises(ValueError, match=r"`filename` cannot contain a path separator."),
+        ),
```

Comment:
Suggest to add a couple more url cases:
```
# url="www.example.net"
>       self.filename = filename or self.url.rsplit("/", 1)[1]
E       IndexError: list index out of range

# url ="http://www.example.net"
=> HTTPSource(url='http://www.example.net', len=0, sha256='', filename='www.example.net')
```

Reply:
Why not parse the `url` with `urllib` and read the filename part from that?
Think it would be slightly more robust by default..
```python
urllib.parse.urlparse("http://www.example.net/demo")
ParseResult(scheme='http', netloc='www.example.net', path='/demo', params='', query='', fragment='')
```